Unify SEH wrapping for method calls in the interpexec.cpp

janvorli opened the pull request:
This change adds a way to wrap any method call in a SEH exception try / catch. Until now, we were adding a new function for each case that needed this wrapping.

I've verified that the generated code in the release build is the same as before this change.

janvorli committed:
Unify SEH wrapping for method calls in the interpexec.cpp

This change adds a templated way to wrap any method call in a SEH
exception try / catch. Until now, we were adding a new function for each
case that needed this wrapping.

jkoritzinsky commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -29,6 +31,46 @@ LONG IgnoreCppExceptionFilter(PEXCEPTION_POINTERS pExceptionInfo, PVOID pv)
+}
+
+template<typename Obj, typename Method, typename... Args>
+auto CallWithSEHWrapper(Obj* obj, Method method, Args... args) -> decltype((obj->*method)(args...))
```
Could we instead use a no-parameter lambda here (and capture the args into the state object)? Would avoid pulling in tuple and that machinery.

jkoritzinsky replied:
One of the reasons I prefer not using the tuple is it violates our C++ guidelines for the repo.

I also think this implementation looks much cleaner and easier to understand at a glance.

janvorli replied:
I don't see anything about tuple violating our C++ guidelines. I assume you may be referring to https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-code-guide.md#-2114-limit-usage-of-standard-template-types-in-shipping-executables. My understanding was that this section was more about stuff that would have parts of the code in the lib. The tuple, at least in this usage case, doesn't require any such code.

We had bad experience with performance of lambdas in the past (@davidwrighton tried to change some huge macros in GC to lambdas), so I was worried that the compiler would not be able to eliminate the calls in this case. Fortunately, it is not that case here.

jkoritzinsky replied:
The concern @jkotas has (and the reason for the rules) is about any code that may exist in the headers or any statically-linked portions. Anything in the standard library that is loaded at runtime from the target machine's standard library implementation is fine.

jkoritzinsky replied:
tuple has code that runs that's defined in the headers, so it meets the definition AFAIK. In practice we'd never have an issue, but my understanding is that we'd rather be more strict here than trying to be more granular header-by-header.

janvorli committed:
Change the implementation to use lambda instead

janvorli commented:
Hmm, the new version breaks the interpreter test, I am investigating ...

janvorli commented:
Ah, copy and paste error for the GetMethodDescOfVirtualizedCode

janvorli committed:
Apply suggestion from @janvorli

Fix copy and paste error

janvorli committed:
Remove CallGetMethodDescOfVirtualizedCode function

Removed the CallGetMethodDescOfVirtualizedCode function, which is now a dead code.

janvorli commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -2453,7 +2416,11 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
-                    targetMethod = CallGetMethodDescOfVirtualizedCode(pMD, pThisArg);
+                    targetMethod = CallWithSEHWrapper(
+                        [&pMD, &pThisArg]() {
+                            return pMD->GetMethodDescOfVirtualizedCode(pMD, pThisArg);
```
```suggestion
                            return pMD->GetMethodDescOfVirtualizedCode(pThisArg, pMD->GetMethodTable());
```

janvorli committed:
Apply suggestion from @janvorli

Fix arguments
